batch/batch/driver/instance_pool.py

```diff
@@ -400,7 +400,10 @@ def add_instance(self, instance):
         while True:
             try:
                 ready_cores = await self.db.select_and_fetchone(
-                    'SELECT * FROM ready_cores;')
+                    '''
+SELECT CAST(SUM(ready_cores_mcpu) AS SIGNED) AS ready_cores_mcpu
```

Comment:
Why the cast?

Reply:
I thought maybe SUM returns a float, but that doesn't seem to be the case:

```
mysql> create temporary table typeof as select SUM(ready_cores_mcpu) from ready_cores;
Query OK, 1 row affected (0.00 sec)
Records: 1  Duplicates: 0  Warnings: 0

mysql> describe typeof;
+-----------------------+---------------+------+-----+---------+-------+
| Field                 | Type          | Null | Key | Default | Extra |
+-----------------------+---------------+------+-----+---------+-------+
| SUM(ready_cores_mcpu) | decimal(32,0) | YES  |     | NULL    |       |
+-----------------------+---------------+------+-----+---------+-------+
1 row in set (0.01 sec)
```

(Note, I'm just summing over one row here, this is with the current schema.)

Reply:
In Python, that value becomes decimal.Decimal() without the cast. Can’t do math between floats and ints and decimals

Reply:
OK, sounds good.  Thanks for clarifying!

Reply:
> AS SIGNED

I have one more question about this.  What's the precision of signed?  Or how is the precision of the result determined?  In light of the INT => BIGINT change, I'm worried about potential overflow.

Reply:
I did some experiments. I think we need to handle the decimal coming out of mysql in Python. I'm still looking what the best way to do that without truncating values by casting to an int in Python (or maybe that does work).

```
mysql> select * from user_resources;
+--------+-------+--------------+------------------+
| user   | token | n_ready_jobs | ready_cores_mcpu |
+--------+-------+--------------+------------------+
| cseed  |     0 |           10 |                1 |
| cseed  |     1 |          -10 |               -1 |
| cseed  |     2 |            5 |                5 |
| jigold |     0 |            3 |                5 |
| jigold |     1 |           -2 |               -3 |
| jigold |     2 |            5 |               10 |
+--------+-------+--------------+------------------+
6 rows in set (0.00 sec)

mysql> desc user_resources;
+------------------+--------------+------+-----+---------+-------+
| Field            | Type         | Null | Key | Default | Extra |
+------------------+--------------+------+-----+---------+-------+
| user             | varchar(100) | NO   | PRI | NULL    |       |
| token            | int(11)      | NO   | PRI | 0       |       |
| n_ready_jobs     | int(11)      | NO   |     | 0       |       |
| ready_cores_mcpu | bigint(20)   | NO   |     | 0       |       |
+------------------+--------------+------+-----+---------+-------+
4 rows in set (0.00 sec)

mysql> select SUM(ready_cores_mcpu) from user_resources;
+-----------------------+
| SUM(ready_cores_mcpu) |
+-----------------------+
|                    17 |
+-----------------------+
1 row in set (0.00 sec)

mysql> CREATE TABLE agg select SUM(ready_cores_mcpu) from user_resources;
Query OK, 1 row affected (0.04 sec)
Records: 1  Duplicates: 0  Warnings: 0

mysql> desc agg
    -> ;
+-----------------------+---------------+------+-----+---------+-------+
| Field                 | Type          | Null | Key | Default | Extra |
+-----------------------+---------------+------+-----+---------+-------+
| SUM(ready_cores_mcpu) | decimal(41,0) | YES  |     | NULL    |       |
+-----------------------+---------------+------+-----+---------+-------+
1 row in set (0.01 sec)

mysql> select * from agg;
+-----------------------+
| SUM(ready_cores_mcpu) |
+-----------------------+
|                    17 |
+-----------------------+
1 row in set (0.00 sec)

mysql> CREATE TABLE agg2 select CAST(SUM(ready_cores_mcpu) AS SIGNED) from user_resources;
Query OK, 1 row affected (0.05 sec)
Records: 1  Duplicates: 0  Warnings: 0

mysql> select * from agg2;
+---------------------------------------+
| CAST(SUM(ready_cores_mcpu) AS SIGNED) |
+---------------------------------------+
|                                    17 |
+---------------------------------------+
1 row in set (0.00 sec)

mysql> desc agg2;
+---------------------------------------+------------+------+-----+---------+-------+
| Field                                 | Type       | Null | Key | Default | Extra |
+---------------------------------------+------------+------+-----+---------+-------+
| CAST(SUM(ready_cores_mcpu) AS SIGNED) | bigint(21) | YES  |     | NULL    |       |
+---------------------------------------+------------+------+-----+---------+-------+
1 row in set (0.01 sec)
```

Reply:
I think we can just do int(x) if x is a decimal.Decimal() and it will work.

```python3
>>> import decimal
>>> bignumber = 1.6 * 10**100
>>> bignumber
1.6000000000000001e+100
>>> int(bignumber)
16000000000000001420047593091596131301054437796423794209094227523680254786560055540644123100496003072
>>> decimal.Decimal(16000000000000001420047593091596131301054437796423794209094227523680254786560055540644123100496003072)
Decimal('16000000000000001420047593091596131301054437796423794209094227523680254786560055540644123100496003072')
>>> x = decimal.Decimal(16000000000000001420047593091596131301054437796423794209094227523680254786560055540644123100496003072)
>>> int(x)
16000000000000001420047593091596131301054437796423794209094227523680254786560055540644123100496003072
```

Reply:
> CAST(SUM(ready_cores_mcpu) AS SIGNED) | bigint(21)

It looks like AS SIGNED casts to BIGINT.  Casting in Python also seems fine.  Both work for me, whatever you prefer.

Reply:
I think this sum needs a COALESCE now, and a couple other places.  Just double-check all the SUMs.